### PR TITLE
feat:  customize name of Shield Tables

### PR DESCRIPTION
**Pull request by datamweb**

### CODE REVIEW

**Comment by datamweb** [comment]
What has happened PHPStan! There is no error in local!

Screenshot 2023-02-06 212606

And PHPUNIT test !

Screenshot 2023-02-06 213030

File: src/Config/Auth.php [hunk 1]
```diff
@@ -37,6 +37,36 @@ class Auth extends BaseConfig
...
+     */
+    public array $authTables = [
+        'users'                  => 'users',
+        'auth_identities'        => 'auth_identities',
+        'auth_logins'            => 'auth_logins',
+        'auth_token_logins'      => 'auth_token_logins',
+        'auth_remember_tokens'   => 'auth_remember_tokens',
+        'auth_groups_users'      => 'auth_groups_users',
+        'auth_permissions_users' => 'auth_permissions_users',
+    ];
```
Comment (kenjis) [inline src/Config/Auth.php:68]:
```suggestion
    public array $tables = [
        'users'             => 'users',
        'identities'        => 'auth_identities',
        'logins'            => 'auth_logins',
        'token_logins'      => 'auth_token_logins',
        'remember_tokens'   => 'auth_remember_tokens',
        'groups_users'      => 'auth_groups_users',
        'permissions_users' => 'auth_permissions_users',
    ];
```

**Comment by kenjis** [comment]
> What has happened PHPStan! There is no error in local!

I also have no error on my local environment.
But `config()` returns `object|null`. 
So `null->authTables` is "Trying to get property 'authTables' of non-object".

**Comment by kenjis** [comment]
@datamweb 
> And PHPUNIT test !

The test does not seem to pass on Windows from the beginning.

`CLI::write()` outputs `PHP_EOL`, but the test code uses `LF` as newline code.

File: src/Config/Auth.php (hunk 1 above)
Comment (datamweb) [reply src/Config/Auth.php]: Done.

**Comment by datamweb** [comment]
@kenjis , thanks for the explanation, it made sense.
  But how can I fix error PHPStan?

File: src/Config/Constants.php [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * Constants Name of Shield Tables
+ */
+define('SHIELD_TABLES', config('Auth')->tables);
```
Comment (kenjis) [inline src/Config/Constants.php:8]: How about this?
```suggestion
/** @var \CodeIgniter\Shield\Config\Auth authConfig */
$authConfig = config('Auth');
define('SHIELD_TABLES', $authConfig->tables);
```

**Comment by datamweb** [comment]
I thought it could be fixed.
I'm surprised, even when I tried to pass the check with `@phpstan-ignore-line` the error was still detected.

File: src/Config/Constants.php [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * Constants Name of Shield Tables
+ */
+/** @var \CodeIgniter\Shield\Config\Auth authConfig */
```
Comment (kenjis) [inline src/Config/Constants.php:8]: Sorry, there was no `$`.
```suggestion
/** @var \CodeIgniter\Shield\Config\Auth $authConfig */
```

**Comment by kenjis** [comment]
Why?

```
Error: Attempt to read property "tables" on null
 ------ ------------------------------------------- 
  Line   src/Config/Constants.php                   
 ------ ------------------------------------------- 
  10     Attempt to read property "tables" on null  
 ------ ------------------------------------------- 
```

**Comment by datamweb** [comment]
**PHP 7.4:**
```console 
Error: Trying to get property 'tables' of non-object
 ------ ----------------------------------------------- 
  Line   src/Config/Constants.php                       
 ------ ----------------------------------------------- 
  10     Trying to get property 'tables' of non-object  
 ------ -----------------------------------------------
```

**Comment by datamweb** [comment]
@kenjis , after many efforts, I finally found the solution.
But I don't know exactly where the main problem is, I think it's in the framework , please let me know if you understand something.

**Comment by kenjis** [comment]
@datamweb I don't know what's wrong with PHPStan.
To begin with, I cannot reproduce the error.

But why don't we remove the constant?
I think using global constant is not good practice especially in libraries.
Global constants are for a user app, and a library should not consume them.

Ion Auth also havs $table, but it does not use a constant.
https://github.com/benedmunds/CodeIgniter-Ion-Auth/blob/6ebcd542011dbc35045466acad0172d2b68ae051/Config/IonAuth.php#L49-L54

We can get the values with `config('Auth')->tables`.

**Comment by datamweb** [comment]
> But why don't we remove the constant?

The reason I used **constant** was to add this feature with minimal changes.(
const
) vs 
(wo_const)

Also, I think it is easier to read and review with this implementation. For example, look at PR [commit](https://github.com/codeigniter4/shield/pull/617/commits/85180598ca3b4e9b42f3cd58f2187d8db2be7f85#diff-74b8b9ef03e167dc3bc070a014429d38eeea0506f27edd419af96f658be1685aR10), it's easier to suggest a change for `SHIELD_TABLES['users']` or 
```php 
    /**
     * Auth Table names
     */
    private array $tables;

    public function __construct(?Forge $forge = null)
    {
        parent::__construct($forge);

        /** @var Auth $authConfig */
        $authConfig   = config('Auth');
        $this->tables = $authConfig->tables;
    }
// and ...

```
In fact, I believe that using **constant** makes our codes shorter and more readable.
Also I was going to include a constant for the shield version, which now I don't know where it fits with the no **constant** implementation.

Technically yes you are right, but we have an example of using **constant** in Bonfire2.

File: src/Config/Registrar.php [hunk 4]
```diff
@@ -14,6 +14,8 @@
 use CodeIgniter\Shield\Filters\SessionAuth;
 use CodeIgniter\Shield\Filters\TokenAuth;
 
+include_once __DIR__ . '/Constants.php';
```
Comment (kenjis) [inline src/Config/Registrar.php:17]: I don't know we should use the constant or not.
But even if we use it, I would like to remove the include_once line and the constant file.

**Comment by kenjis** [comment]
If we add `SHIELD_VERSION`, we should add it as a class constant like `CodeIgniter\CodeIgniter::CI_VERSION`.
Class constants do not consume the global namespace.

**Comment by datamweb** [comment]
Okay, we will continue in your PR.

**Pull request closed without merging**